### ISSUE TITLE
set tcp options

### ISSUE THREAD

**Issue by jnordwick** [snapview/tungstenite-rs#277]
I need to set some tcp options on a tcp socket. Nagel off, and some Linux flow control off too. Two of these tcp send and receive buffer sizes should also be known to whatever is doing the actual reading so it can size its buffers too (also, those two need to be done before connect is called so tcp window size an be set properly). I probably set the recv to ~4MB or larger. The send I'll bump to at least 64K.

Kind of related, how do I do reconnects? Since it also needs to know about this stuff. (using tokio, but tokio says socket creates is outside its scope)

**Comment by application-developer-DA** [comment]
> I need to set some tcp options on a tcp socket. Nagel off, and some Linux flow control off too. Two of these tcp send and receive buffer sizes should also be known to whatever is doing the actual reading so it can size its buffers too (also, those two need to be done before connect is called so tcp window size an be set properly). I probably set the recv to ~4MB or larger. The send I'll bump to at least 64K.

You can e.g. create a `TcpStream`, set any options and pass [it to the function in `tungstenite`](https://docs.rs/tungstenite/latest/tungstenite/client/fn.client.html). E.g. configuring of TCP specific option and other logic is out of scope for `tungstenite` (has nothing to do with the WebSockets), we are stream-agnostic, i.e. you can modify the `TcpStream` (of in fact any `Read + Write` stream) and pass it to the `tungstenite` / `tokio-tungstenite`, we'll just work on top of what we received from the user.

> Kind of related, how do I do reconnects? Since it also needs to know about this stuff. (using tokio, but tokio says socket creates is outside its scope)

A reconnect == a reestablishment of the connection 🙂 If the `WebSocket` connection gets closed, you can just establish a new one (both in `tungstenite` and `tokio-tungstenite`).

Hope this helps!

**Comment by application-developer-DA** [comment]
Accepting connection is also not a problem since it [also receives a stream](https://docs.rs/tungstenite/latest/tungstenite/fn.accept.html), for even more control, the one can even call the function directly [on the `WebSocket` level](https://docs.rs/tungstenite/latest/tungstenite/protocol/struct.WebSocket.html#method.from_raw_socket).

> Does this mean I would need to do my own SNI stuff for TLS? And I thik the quck connect stuff also needs support on the connection calls too, but I 'm not sure fully how that is controlled.

Depends on what you want to achieve. Both `tungstenite` and Tokio are quite flexible in the way. You can use the helpers from `tungstenite` and call to the `connect` with your own `Connector` or you can leave this part out if you don't need it. It all really depends on what the user wants, we provide the "core" (lower-level stuff that deals with WebSockets) and also a couple of helpers for the most typical cases that cover maybe 90% of cases.

**Comment by jnordwick** [comment]
Im not sure what the streams has to do with anything, but my point was that the listening socket needs to have some options set on it, when accept completes, the descriptor needs to be fished out and other options need to be set before the anything else can continue - not all socket options are inheritable and some are required to be set early on so they can be used. And some values need to be communicated to other parts of the system (eg, tcp buffer sizes for TLS) -  a very specific thing would be how do i set the recv buffer for a TLS server socket?

The Connectors looks usable, but tcp socket creation would have to be tied through it (the code I saw  had TCPSocekt::connect being called before Connector::connect, and those should be the same  if values like buf size are to be pulled from it, and some values like resuse port would have to be set too. UDP has a another list of these. 

I still have find out how to change internal buffer sizes to align with the kernel values. For most high perforamcne non-blocking socket code this is common - as well as disabling certain tcp flow controls.

Would it be preferable to have a stack of these connnectors? or maybe just throw all the values in a single Connector and let each layer pick through them? And on top of the tranport protocols, there will are very similar values for the application level value too.. eg- max datagram size, nack window, etc...
